astropy/table/_dataframes.py

```diff
@@ -0,0 +1,517 @@
+) -> "Table":
+    """Create a Table from any narwhals-compatible DataFrame."""
+    from .table import Table
+
+    if not HAS_NARWHALS:
+        raise ModuleNotFoundError(
+            "The narwhals library is required for the generic from_df method. "
+            "If you want to only convert from pandas, use the `from_pandas` method instead."
+        )
+    import narwhals as nw
+
+    # Create output
+    out = OrderedDict()
+
+    # Handle pandas index
+    if index:
+        if not hasattr(df, "index"):
+            raise ValueError(
+                "The input dataframe does not have an index. "
+                "Are you trying to convert a non-pandas dataframe? "
+                "Set `index=False` to avoid this error."
+            )
+        index_name = df.index.name or "index"
+        while index_name in df.columns:
+            index_name = "_" + index_name + "_"
+        df.reset_index(index_name, inplace=True, drop=False)
+
+    # Narwhals layer, must convert to eager
+    df_nw = nw.from_native(df)
+    if isinstance(df_nw, nw.LazyFrame):
+        df_nw = df_nw.collect()
+
+    # Handle units
+    if units is None:
+        units = {}
+    elif not isinstance(units, Mapping):
+        raise TypeError('Expected a Mapping "column-name" -> "unit"')
+    not_found = set(units.keys()) - set(df_nw.columns)
+    if not_found:
+        warnings.warn(f"`units` contains additional columns: {not_found}")
+
+    # Iterate over Narwhals columns
+    for column in df_nw.iter_columns():
+        # Unpack relevant data
+        data = column.to_numpy()
+        name = column.name
+        dtype = column.dtype
+        mask = column.is_null().to_numpy()
+        unit = units.get(name)
+
+        if isinstance(dtype, nw.Int128):
+            raise ValueError(
+                "Numpy and astropy Table do not support Int128, please use a smaller integer type."
+            )
+
+        # Handle nullable integers
+        if dtype.is_integer() and mask.any():
+            nullfilled = column.fill_null(0).to_numpy()
+            out[name] = MaskedColumn(data=nullfilled, mask=mask, unit=unit, copy=False)
+            continue
+
+        # Handle datetime columns
+        elif isinstance(dtype, nw.Datetime):
+            from astropy.time import Time
+
+            datetime = Time(data, format="datetime64")
+            datetime.format = "isot"
+            out[name] = datetime
+            continue
+
+        # Handle timedelta columns
+        elif isinstance(dtype, nw.Duration):
+            from astropy.time import TimeDelta
+
+            duration = data.astype("timedelta64[ns]").astype(np.float64) / 1e9
+            out[name] = TimeDelta(duration, format="sec")
+            if mask.any():
+                out[name][mask] = np.ma.masked
+            continue
+
+        # Handle string-like columns
+        elif isinstance(dtype, (nw.String, nw.Binary, nw.Object)):
```

Comment:
This one is a shot in the dark. Please ignore if it doesn't make sense
```suggestion
        elif not dtype.is_numeric():
```

Reply:
> The question is are there non-numeric types on which we don`t want to perform these string corrections?

You can check here for which types are considered numeric https://narwhals-dev.github.io/narwhals/api-reference/dtypes/

Possibly `nw.Enum` and `nw.Categorical` would be the other non-numeric types of interest

Reply:
makes sense, I didn't have much faith in my suggestion any way :)